core/src/main/java/jenkins/install/StartupUtil.java

```diff
@@ -0,0 +1,159 @@
+    }
+
+    /**
+     * Get the last saved Jenkins instance version.
+     * @return The last saved Jenkins instance version.
+     * @see #saveLastExecVersion()
+     */
+    public static @Nonnull String getLastExecVersion() {
+        File lastExecVersionFile = getLastExecVersionFile();
+        if (lastExecVersionFile.exists()) {
+            try {
+                return FileUtils.readFileToString(lastExecVersionFile);
+            } catch (IOException e) {
+                throw new IllegalStateException("Unexpected Error. Unable to read " + lastExecVersionFile.getAbsolutePath(), e);
+            }
+        } else {
+            // Backward compatibility. Use the last version stored in the top level config.xml.
+            // Going to read the value directly from the config.xml file Vs hoping that the 
+            // Jenkins startup sequence has moved far enough along that it has loaded the
+            // global config. It can't load the global config until well into the startup
+            // sequence because the unmarshal requires numerous objects to be created e.g.
+            // it requires the Plugin Manager. It happens too late and it's too risky to
+            // change how it currently works.
+            File configFile = getConfigFile();
+            if (configFile.exists()) {
+                try {
+                    String lastVersion = XMLUtils.getValue("/hudson/version", configFile);
+                    if (lastVersion.length() > 0) {
+                        return lastVersion;
+                    }
+                } catch (Exception e) {
+                    LOGGER.log(SEVERE, "Unexpected error reading global config.xml", e);
+                }
+            }
+            return NEW_INSTALL_VERSION.toString();
+        }
+    }
+
+    /**
+     * Save a specific version as the last execute version.
+     * @param version The version to save.
+     */
+    static void saveLastExecVersion(@Nonnull String version) {
+        File lastExecVersionFile = getLastExecVersionFile();
+        try {
+            FileUtils.write(lastExecVersionFile, version);
+        } catch (IOException e) {
+            LOGGER.log(SEVERE, "Failed to save " + lastExecVersionFile.getAbsolutePath(), e);
+        }
+    }
+
+    static File getConfigFile() {
+        return new File(Jenkins.getActiveInstance().getRootDir(), "config.xml");
+    }
+
+    static File getLastExecVersionFile() {
+        return new File(Jenkins.getActiveInstance().getRootDir(), ".last_exec_version");
+    }
+
+    private static String getCurrentExecVersion() {
+        if (Jenkins.VERSION.equals(Jenkins.UNCOMPUTED_VERSION)) {
+            // This should never happen!! Only adding this check in case someone moves the call to this method to the wrong place.
+            throw new IllegalStateException("Unexpected call to StartupUtil.getCurrentExecVersion(). Jenkins.VERSION has not been initialized. Call computeVersion() first.");
```

Comment:
Should rather make all of `StartupUtil` be `@Restricted(NoExternalUse.class)` so we know for sure how it is called.